rust/lance/src/dataset/sql.rs

```diff
@@ -0,0 +1,214 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: Copyright The Lance Authors
+
+use crate::datafusion::LanceTableProvider;
+use crate::Dataset;
+use arrow_array::{Array, RecordBatch, StringArray};
+use datafusion::dataframe::DataFrame;
+use datafusion::execution::SendableRecordBatchStream;
+use datafusion::prelude::SessionContext;
+use std::sync::Arc;
+
+/// A SQL builder to prepare options for running SQL queries against a Lance dataset.
+#[derive(Clone, Debug)]
+pub struct SqlBuilder {
+    /// The dataset to run the SQL query
+    pub(crate) dataset: Dataset,
+
+    /// The SQL query to run
+    pub(crate) sql: String,
+
+    /// the name of the table to register in the datafusion context
+    pub(crate) table_name: Option<String>,
+
+    /// If true, the query result will include the internal row id
+    pub(crate) row_id: Option<bool>,
```

Comment:
should be named `with_row_id` and `with_row_addr` to match the inputs for `LanceTableProvider`